provider/google/metadata_test.go

```diff
@@ -0,0 +1,98 @@
+package google
+
+import (
+	"fmt"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/taskcluster/httpbackoff"
+)
+
+type fakeMetadataService struct {
+	UserDataError error
+	UserData      *UserData
+	Metadata      map[string]string
+}
+
+func (mds *fakeMetadataService) queryUserData() (*UserData, error) {
+	if mds.UserDataError != nil {
+		return nil, mds.UserDataError
+	}
+	return mds.UserData, nil
+}
+
+func (mds *fakeMetadataService) queryMetadata(path string) (string, error) {
+	if path[0] != '/' {
+		panic("path must start with /")
+	}
+	res, ok := mds.Metadata[path]
+	if !ok {
+		return "", fmt.Errorf("not found: %s", path)
+	}
+	return res, nil
+}
+
+func TestQueryMetadata(t *testing.T) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Header.Get("Metadata-Flavor") != "Google" {
+			w.WriteHeader(400)
+			fmt.Fprintln(w, "Metadata-Flavor Missing")
+		} else if r.URL.Path == "/computeMetadata/v1/project/id" {
+			w.WriteHeader(200)
+			fmt.Fprintln(w, "42")
+		} else {
+			w.WriteHeader(404)
+			fmt.Fprintln(w, "Not Found")
+		}
+	}))
+	defer ts.Close()
+
+	metadataBaseURL = ts.URL + "/computeMetadata/v1/project"
+	defer func() {
+		metadataBaseURL = "http://metadata.google.internal/computeMetadata/v1/project"
+	}()
+
+	ms := realMetadataService{}
+
+	rv, err := ms.queryMetadata("/id")
+	if assert.NoError(t, err) {
+		assert.Equal(t, "42\n", rv)
+	}
+
+	_, err = ms.queryMetadata("/meta-data/NOSUCH")
+	if assert.Error(t, err) {
+		httperr, ok := err.(httpbackoff.BadHttpResponseCode)
+		assert.True(t, ok)
+		assert.Equal(t, 404, httperr.HttpResponseCode)
+	}
+}
+
+func TestQueryUserData(t *testing.T) {
+	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.Header.Get("Metadata-Flavor") != "Google" {
+			w.WriteHeader(400)
+			fmt.Fprintln(w, "Metadata-Flavor Missing")
+		} else if r.URL.Path == "/computeMetadata/v1/project/instance/attributes/taskcluster" {
+			w.WriteHeader(200)
+			fmt.Fprintln(w, `{"workerPoolId": "w/p"}`)
+		} else {
+			w.WriteHeader(404)
+			fmt.Fprintf(w, "Not Found: %s", r.URL.Path)
```

Comment:
Is the reason this is different between the two tests that you're emulating the real metadata service responses?

Reply:
Yes.  I probably could use a single httptest.Server too, but this works.